runners/sbt/src/main/scala/stryker4s/sbt/SbtMutantRunner.scala

```diff
@@ -1,73 +1,68 @@
-    Project.runTask(test in Test, newState) match {
-      case None =>
-        throw InitialTestRunFailedException(
-          s"Unable to execute initial test run. Sbt is unable to find the task 'test'.")
-      case Some((_, Value(_))) => true
-      case Some((_, Inc(_)))   => false
-    }
-  }
+  private val extracted = Project.extract(state)
 
-  override def runMutant(mutant: Mutant, workingDir: File, subPath: Path): MutantRunResult = {
-    val newState = extracted.appendWithoutSession(settings(workingDir) ++ mutationSetting(mutant.id), state)
+  private val newState = extracted.appendWithoutSession(settings, state)
 
-    Project.runTask(test in Test, newState) match {
-      case None =>
-        throw new RuntimeException(s"An unexpected error occurred while running mutation ${mutant.id}")
-      case Some((_, Value(_))) => Survived(mutant, subPath)
-      case Some((_, Inc(_)))   => Killed(mutant, subPath)
-    }
+  override def runInitialTest(workingDir: File): Boolean = runTests(
+    newState,
+    InitialTestRunFailedException(s"Unable to execute initial test run. Sbt is unable to find the task 'test'."),
+    onSuccess = true,
+    onFailed = false
+  )
+
+  override def runMutant(mutant: Mutant, workingDir: File, subPath: Path): MutantRunResult = {
+    val mutationState = extracted.appendWithSession(settings :+ mutationSetting(mutant.id), newState)
+    runTests(
+      mutationState,
+      MutationRunFailedException(s"An unexpected error occurred while running mutation ${mutant.id}"),
+      Survived(mutant, subPath),
+      Killed(mutant, subPath)
+    )
   }
 
-  private[this] def settings(tmpDir: File): Seq[Def.Setting[_]] = {
-    val mainPath = {
-      extracted
-        .get(Compile / scalaSource)
-        .absolutePath
-        .diff(
-          extracted.get(Compile / baseDirectory).absolutePath
-        )
+  /** Runs tests with the giving state, calls the corresponding parameter on each result
+    */
+  private def runTests[T](state: State, onError: => Stryker4sException, onSuccess: => T, onFailed: => T): T =
+    Project.runTask(test in Test, state) match {
+      case None                => throw onError
+      case Some((_, Value(_))) => onSuccess
+      case Some((_, Inc(_)))   => onFailed
     }
 
-    val testPath = {
-      extracted
-        .get(Test / scalaSource)
-        .absolutePath
-        .diff(
-          extracted.get(Test / baseDirectory).absolutePath
-        )
-    }
+  private def mutationSetting(mutation: Int): Def.Setting[_] =
+    javaOptions in Test += s"-DACTIVE_MUTATION=${String.valueOf(mutation)}"
 
-    Seq(
-      fork in Test := true,
-      scalaSource in Compile := tmpDir.toJava / mainPath,
-      scalaSource in Test := tmpDir.toJava / testPath
-    )
+  private def tmpDirFor(conf: Configuration): Def.Initialize[JFile] = {
+    val sourceDef = pathOf(scalaSource, conf)
+    val baseDef = pathOf(baseDirectory, conf)
 
+    (sourceDef zipWith baseDef)(_ diff _)(tmpDir.toJava / _)
```

Comment:
I'm having a hard time understanding what exactly happens here. Fancy oneliners are cool but I think this could do with a bit more readability. 😀

Reply:
Yeah this is better. Fewer underscores helps. :)